common/src/main/java/tachyon/resource/ResourcePool.java

```diff
@@ -66,19 +67,33 @@ protected ResourcePool(int maxCapacity, BlockingQueue<T> resources) {
    */
   public T acquire() {
     // If the resource pool is empty but capacity is not yet full, create a new resource.
-    synchronized (mCapacityLock) {
-      if (mResources.isEmpty() && mCurrentCapacity < mMaxCapacity) {
-        T newResource = createNewResource();
-        mCurrentCapacity ++;
-        return newResource;
-      }
+    T resource = mResources.poll();
+    if (resource != null) {
+      return resource;
+    }
+
+    if (mCurrentCapacity.getAndIncrement() < mMaxCapacity) {
+      return createNewResource();
     }
 
+    mCurrentCapacity.decrementAndGet();
+
     // Otherwise, try to take a resource from the pool, blocking if none are available.
     try {
-      return mResources.take();
+      mTakeLock.lockInterruptibly();
+      try {
+        while (true) {
+          resource = mResources.poll();
+          if (resource != null) {
+            mNotEmpty.signal();
```

Comment:
That's right, this line is not mandatory and release()'s signal() is surely enough. 

release() add an entry to the queue then send signal. Because this line is already in the lock, it could send signal before release() send signal. release()'s signal() would do nothing in that case. But basically, it must be very very trivial for the performance or could be worse. So I'm going to simply remove it.